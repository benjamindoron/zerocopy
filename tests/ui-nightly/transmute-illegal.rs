// Copyright 2022 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

extern crate zerocopy;

fn main() {}

// It is unsound to inspect the usize value of a pointer during const eval.
const POINTER_VALUE: usize = zerocopy::transmute!(&0usize as *const usize);

// `transmute!` and `try_transmute!` enforce size equality.
const TOO_LARGE: u64 = zerocopy::transmute!(0u8);
const TRY_TOO_LARGE: Option<u64> = zerocopy::try_transmute!(0u8);
const TOO_SMALL: u8 = zerocopy::transmute!(0u64);
const TRY_TOO_SMALL: Option<u8> = zerocopy::try_transmute!(0u64);
