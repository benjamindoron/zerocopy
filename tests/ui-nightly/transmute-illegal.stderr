error[E0277]: the trait bound `*const usize: AsBytes` is not satisfied
  --> tests/ui-nightly/transmute-illegal.rs:10:30
   |
10 | const POINTER_VALUE: usize = zerocopy::transmute!(&0usize as *const usize);
   |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                              |
   |                              the trait `AsBytes` is not implemented for `*const usize`
   |                              required by a bound introduced by this call
   |
   = help: the trait `AsBytes` is implemented for `usize`
note: required by a bound in `POINTER_VALUE::transmute`
  --> tests/ui-nightly/transmute-illegal.rs:10:30
   |
10 | const POINTER_VALUE: usize = zerocopy::transmute!(&0usize as *const usize);
   |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `transmute`
   = note: this error originates in the macro `zerocopy::transmute` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
  --> tests/ui-nightly/transmute-illegal.rs:13:24
   |
13 | const TOO_LARGE: u64 = zerocopy::transmute!(0u8);
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: source type: `u8` (8 bits)
   = note: target type: `u64` (64 bits)
   = note: this error originates in the macro `zerocopy::transmute` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
  --> tests/ui-nightly/transmute-illegal.rs:14:36
   |
14 | const TRY_TOO_LARGE: Option<u64> = zerocopy::try_transmute!(0u8);
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: source type: `u8` (8 bits)
   = note: target type: `u64` (64 bits)
   = note: this error originates in the macro `zerocopy::try_transmute` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
  --> tests/ui-nightly/transmute-illegal.rs:15:23
   |
15 | const TOO_SMALL: u8 = zerocopy::transmute!(0u64);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: source type: `u64` (64 bits)
   = note: target type: `u8` (8 bits)
   = note: this error originates in the macro `zerocopy::transmute` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
  --> tests/ui-nightly/transmute-illegal.rs:16:35
   |
16 | const TRY_TOO_SMALL: Option<u8> = zerocopy::try_transmute!(0u64);
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: source type: `u64` (64 bits)
   = note: target type: `u8` (8 bits)
   = note: this error originates in the macro `zerocopy::try_transmute` (in Nightly builds, run with -Z macro-backtrace for more info)
